src/main/scala/mesosphere/marathon/core/storage/PersistenceStore.scala

```diff
@@ -0,0 +1,163 @@
+    um: Unmarshaller[Serialized, V]): Future[Option[V]]
+
+  /**
+    * Get the version of the data at the given version, if any, for the given primary id and value type.
+    *
+    * @return A future representing the data at the given Id, if any exists.
+    *         If there is an underlying storage problem, the future should fail with
+    *         [[mesosphere.marathon.StoreCommandFailedException]]
+    */
+  def get[Id, V](
+    id: Id,
+    version: OffsetDateTime)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    um: Unmarshaller[Serialized, V]): Future[Option[V]]
+
+  /**
+    * Store the new value at the given Id. If the value already exists, the existing value will be versioned
+    *
+    * @return A Future that will complete with the previous version of the value if it existed, or fail with
+    *         [[mesosphere.marathon.StoreCommandFailedException]] if either a value already exists at the given Id, or
+    *         if there is an underlying storage problem
+    */
+  def store[Id, V](id: Id, v: V)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    m: Marshaller[V, Serialized],
+    um: Unmarshaller[Serialized, V]): Future[Done]
+
+  /**
+    * Store a new value at the given version. If the maximum number of versions has been reached,
+    * will delete the oldest versions. This method does not replace the current version.
+    *
+    * @return A Future that will complete with the previous version of the value if it existed, or fail with
+    *         [[mesosphere.marathon.StoreCommandFailedException]] if either a value already exists at the given Id, or
+    *         if there is an underlying storage problem
+    */
+  def store[Id, V](id: Id, v: V, version: OffsetDateTime)(
+    implicit
+    ir: IdResolver[Id, V, Category, K],
+    m: Marshaller[V, Serialized]): Future[Done]
+
+  /**
+    * Delete the value at the given Id and version, idempotent
+    *
+    * @return A future indicating whether the value was deleted (or simply didn't exist). Underlying storage issues
+    *         will fail the future with [[mesosphere.marathon.StoreCommandFailedException]]
+    */
+  def delete[Id, V](k: Id, version: OffsetDateTime)(implicit ir: IdResolver[Id, V, Category, K]): Future[Done]
+
+  /**
+    * Delete all of the versions of the given Id, idempotent
+    * @return A future indicating whether the value was deleted (or simply didn't exist). Underlying storage issues
+    *         will fail the future with [[mesosphere.marathon.StoreCommandFailedException]]
+    */
+  def deleteAll[Id, V](k: Id)(implicit ir: IdResolver[Id, V, Category, K]): Future[Done]
+
+  /**
+    * @return A source of _all_ keys in the Persistence Store (which can be used by a
+    *         [[mesosphere.marathon.core.storage.impl.cache.LoadTimeCachingPersistenceStore]] to populate the
+    *         cache completely on startup.
+    */
+  protected[storage] def keys(): Source[CategorizedKey[Category, K], NotUsed]
```

Comment:
I still need to move this one. Ids() is actually idsByCategory(). Keys() is all of the ids. That is confusing. This method is exclusively used by the load time cache.